text/0000-asset-loader-service.md

```diff
@@ -0,0 +1,208 @@
+
+```js
+type AssetPromise = Promise<Asset, AssetLoadError>;
+type BundlePromise = Promise<BundleName, BundleLoadError>;
+```
+
+Meaning that an `AssetPromise` will resolve with the information of the `Asset` it was loading while the `BundlePromise` will resolve with the name of the `Bundle` it was loading. In both cases, rejections will be handled by special types of errors which are defined as:
+
+```js
+interface LoadError extends Error {
+  retryLoad(): Promise<T, U>
+}
+
+interface AssetLoadError extends LoadError {
+  asset: Asset,
+  retryLoad(): AssetPromise
+}
+
+interface BundleLoadError extends LoadError {
+  bundle: BundleName,
+  loadErrors: Array<LoadError>,
+  retryLoad(): BundlePromise
+}
+```
+
+The `LoadError` classes will be discussed in-depth in a later section.
+
+## `AssetLoader` Methods
+
+### `pushManifest`
+
+The `pushManifest` method allows configuration of the Asset Loader service by providing an object that conforms to the Asset Manifest specification. This will allow the loader to know what assets are available to load and where to find them.
+
+Additional invocations will add manifests into an internal data structure for later reference. When attempting to load bundles or perform tasks that require referencing the manifests, the `bundles` property in each manifest will be merged. In the case of merge conflicts, an error will be thrown to avoid inconsistent and unintuitive loading behavior.
+
+Since the Asset Loader will not generate a manifest itself, this method is the interface through which to explicitly provide the external manifest to the service.
+
+Note: This method _must_ be called at least once prior to using `loadBundle` in order to provide the information needed to load any bundles. This should likely be done in an instance-initializer.
+
+#### Example
+
+```js
+// instance-initializers/asset-loader-manifest.js
+import manifest from '../config/asset-manifest';
+
+export function initialize(instance) {
+  const loader = instance.lookup('service:asset-loader');
+  loader.pushManifest(manifest);
+}
+
+export default {
+  name: 'asset-loader-manifest',
+  initialize
+};
+```
+
+### `loadAsset`
+
+The `loadAsset` method loads a single asset into the application. It accepts an `asset` object which contains `uri` and `type` fields, both of which are strings. The `type` determines which loading method will be called internally according to the type of asset being loaded and the `uri` will be passed as an argument to said loading method to specify the actual asset to load.
+
+`loadAsset` will return an `AssetPromise` that will resolve when the asset is successfully loaded or reject when the asset fails to load.
+
+Subsequent calls to `loadAsset` with the same arguments will return the same `AssetPromise` object to ensure a single source of truth for determining asset state as well as to prevent duplicate requests being made for the same asset.
```

Comment:
Can we clarify here that this is true unless/until `retryLoad` is called, and link down to the `Provide A Way To Attempt Recovery` section?